CTAA-1389: Risk calculation

finnsch opened the pull request.

finnsch committed:
Merge branch 'story/exposure_framework' into task/CTAA-1389_risk-calculation

finnsch committed:
Merge remote-tracking branch 'origin/story/exposure_framework' into task/CTAA-1389_risk-calculation

finnsch committed:
CTAA-1389: Add first draft for the risk calculation (WIP)

finnsch committed:
CTAA-1389: Add missing import and return

nilsgrabenhorst commented on CoronaContact/Extensions/ENExposureDetectionSummary+lastExposureDate.swift:
```diff
@@ -0,0 +1,13 @@
+
+extension ENExposureDetectionSummary {
+    var lastExposureDate: Date? {
+        Calendar.current.date(byAdding: .day, value: daysSinceLastExposure, to: Date())
```
is `daysSinceLastExposure` a negative value? Otherwise the lastExposureDate is in the future, isn't it?

nilsgrabenhorst replied:
Now fixed, thanks

nilsgrabenhorst commented on CoronaContact/Extensions/ENIntervalNumber+date.swift:
```diff
@@ -0,0 +1,13 @@
+    var date: Date {
+        Date(timeIntervalSince1970: Double(self) * 60 * 10)
+    }
+}
```
The idiomatic way in swift is to add an initialiser to `Date` instead:
```
extension Date {
    init(_ enIntervalNumber: ENIntervalNumber) {
        self.init(timeIntervalSince1970: Double(enIntervalNumber) * 60 * 10)
    }
}
```

finnsch replied:
I am not sure if I am a fan of this. How important is this change for you?

nilsgrabenhorst replied:
Not very important. Just pointing this out. The Standard Library & Foundation mostly use the initialiser pattern, but not always. The exception to the rule that I always trip over is converting String to Data...

nilsgrabenhorst commented on CoronaContact/Logic/RiskCalculation/DetectDailyExposuresOperation.swift:
```diff
@@ -0,0 +1,63 @@
+            if let error = error {
+                self.finish(with: .failure(.exposureDetectionFailed(error)))
+                return
+            }
```
It would be nicer if the closure had a single `Result` argument.

finnsch replied:
Fixed

nilsgrabenhorst commented on CoronaContact/Logic/RiskCalculation/DetectExposuresOperation.swift:
```diff
@@ -0,0 +1,50 @@
+    private let diagnosisKeyURLs: [URL]
+    private var progress: Progress?
+
+    let exposureManager: ExposureManager
```
Could this be `private`?

nilsgrabenhorst replied:
Thanks for fixing.

nilsgrabenhorst commented on CoronaContact/Logic/RiskCalculation/DetectDailyExposuresOperation.swift:
```diff
@@ -0,0 +1,63 @@
+    private let diagnosisKeyURLs: [URL]
+    private var progress: Progress?
+
+    let exposureManager: ExposureManager
```
`private`?

finnsch replied:
Fixed

bastianh commented on CoronaContact/Extensions/ENIntervalNumber+date.swift:
```diff
@@ -0,0 +1,13 @@
+    var date: Date {
+        Date(timeIntervalSince1970: Double(self) * 60 * 10)
+    }
+}
```
können wir das bitte aufsplitten ? ;)
```
extension ENIntervalNumber {
    var timeInterval: TimeInterval {
        Double(self) * 60 * 10
    }

    var date: Date {
        Date(timeIntervalSince1970: timeInterval)
    }
}

finnsch replied:
Fixed :)

nilsgrabenhorst commented on CoronaContact/Logic/RiskCalculation/RiskCalculationController.swift:
```diff
@@ -0,0 +1,71 @@
+        return queue
+    }()
+
+    private let log = ContextLogger(context: .riskCalculation)
```
For some reason I get a compiler error here that can be solved by being more explicit:
```suggestion
    private let log = ContextLogger(context: LoggingContext.riskCalculation)
```

finnsch replied:
Pretty strange that I didn't receive the compiler error. I added the prefix just to be sure.

nilsgrabenhorst commented on CoronaContact/Logic/RiskCalculation/RiskCalculationController.swift:
```diff
@@ -0,0 +1,71 @@
+        return operation
+    }
+
+    private func processDailyBatches(_ batches: [UnzippedBatch], startingFrom date: Date) {
```
`startingFrom` suggests moving forward in time. The second filter suggests that the algorithm actually moves backwards in time. If that's intended I would recommend renaming the argument to this:
`private func processDailyBatches(_ batches: [UnzippedBatch], before date: Date)`

Otherwise double-check the filter.

finnsch replied:
I agree with you. This makes it much clearer.

finnsch committed:
CTAA-1389: Split up ENIntervalNumber extensions

nilsgrabenhorst commented on CoronaContact/Model/ExposureConfiguration.swift:
```diff
@@ -0,0 +1,38 @@
+        exposureConfiguration.metadata = ["attenuationDurationThresholds": attenuationDurationThresholds]
+
+        return exposureConfiguration
+    }
```
I'd prefer a convenience initialiser on `ENExposureConfiguration`

finnsch replied:
Good idea, fixed. :)

finnsch committed:
CTAA-1389: Determine if exposure was a red or yellow exposure

finnsch committed:
CTAA-1389: Return result of risk calculation to the caller

finnsch committed:
Merge branch 'story/exposure_framework' into task/CTAA-1389_risk-calculation

# Conflicts:
#	CoronaContact.xcodeproj/project.pbxproj
#	CoronaContact/Model/Configuration.swift

finnsch committed:
CTAA-1389: Fix calculation of lastExposureDate

finnsch committed:
CTAA-1389: Add support for 7 and 14 day batches

finnsch committed:
CTAA-1389: Update folder structure

finnsch committed:
CTAA-1389: Pass risk calculation result to quarantine time calculation (TODO)

bastianh committed:
Merge branch 'story/exposure_framework' into task/CTAA-1389_risk-calculation

bastianh committed:
removed unused config parameter

bastianh committed:
fix error logging in the configuration service

bastianh committed:
fix last merge

bastianh committed:
store completion handler for later usage

bastianh committed:
calculation of quarantinedays

bastianh committed:
do not remove contact in case of error

bastianh committed:
de-duplicate some code

bastianh committed:
Merge pull request #76 from austrianredcross/task/CTAA-1389-continue

Task/ctaa 1389 some fixes / extensions

bastianh committed:
updated configuration.json

bastianh committed:
localization fix

finnsch committed:
CTAA-1389: Implement suggestions from the code review

finnsch committed:
CTAA-1389: Add more logging for risk calculation

finnsch committed:
CTAA-1389: Add more logging for batch downloads

finnsch committed:
CTAA-1389: Fix configuration.json

finnsch committed:
CTAA-1389: Fix risk detection

Add total risk score to exposure summary
Fix execution order of daily exposure detection operations
Fix bug where the risk calculation result did not include all results

finnsch committed:
CTAA-1389: Disable SwiftFormat for one code line to satisfy compiler

finnsch committed:
CTAA-1389: Store detection interval in a constant

finnsch committed:
CTAA-1389: Fix risk calculation

finnsch committed:
Merge branch 'story/exposure_framework' into task/CTAA-1389_risk-calculation

uliluckas commented on CoronaContact/Extensions/ENExposureDetectionSummary+totalRiskScore.swift:
```diff
@@ -0,0 +1,56 @@
+}
+
+private extension Set where Element == ENRiskLevelValue {
+    mutating func unique() -> UniqueAttenuationLevelValues? {
```
I suggest working on a copy instead of mutating

finnsch replied:
Good idea, I changed the code accordingly. 👍

finnsch committed:
CTAA-1389: Do not mutate the set

finnsch committed:
CTAA-1389: Add private access modifiers where possible

nilsgrabenhorst commented on CoronaContact/Extensions/ENExposureDetectionSummary+totalRiskScore.swift:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        return UniqueAttenuationLevelValues(max: maxValue, min: minValue, median: medianValue)
+    }
```
This method does something else than what the name suggests. It also destroys the value by mutating it, which is not what I would expect. I'd suggest:
* make the method non-mutating
* brainstorm on the method name

```
private extension Set where Element == ENRiskLevelValue {
    func makeUniqueAttenuationValues() -> UniqueAttenuationLevelValues? {
        
        var values = self
        
        values.remove(0)

        guard let maxValue = values.max(), let minValue = values.min() else {
            return nil
        }

        values.remove(maxValue)
        values.remove(minValue)

        guard let medianValue = values.first else {
            return nil
        }

        return UniqueAttenuationLevelValues(max: maxValue, min: minValue, median: medianValue)
    }
}
```

nilsgrabenhorst commented on CoronaContact/Extensions/ENExposureDetectionSummary+totalRiskScore.swift:
```diff
@@ -0,0 +1,56 @@
+        }
+
+        // 5 minutes in seconds
+        let detectionInterval: UInt = 300
```
You can also write 
```suggestion
        let detectionInterval: UInt = 5 * 60
```
The compiler will create a constant for you.

nilsgrabenhorst commented on CoronaContact/Extensions/ENExposureDetectionSummary+totalRiskScore.swift:
```diff
@@ -0,0 +1,56 @@
+
+        guard let medianValue = first else {
+            return nil
+        }
```
Are there only three values (excluding 0)?

finnsch replied:
Yes, Uli has confirmed this. The current approach is rather fragile. Ideally the backend should return only these values as part of the exposure configuration, but this will probably take some time to implement.

nilsgrabenhorst replied:
Maybe add a comment for the casual consumer of this codebase?

finnsch committed:
CTAA-1389: Update user explanation for the call to get exposure info

nilsgrabenhorst commented on CoronaContact/Services/RiskCalculation/RiskCalculationController.swift:
```diff
@@ -69,14 +69,18 @@ final class RiskCalculationController {
         let operations: [DetectDailyExposuresOperation] = dailyBatches.map { batch in
             let operation = DetectDailyExposuresOperation(diagnosisKeyURLs: batch.urls, date: batch.interval.date)
             operation.completionBlock = handleCompletion(of: operation, date: batch.interval.date)
+            operation.handleDailyExposure = { [weak self] in self?.storeDailyExposure($0, at: $1) }
```
```suggestion
            operation.handleDailyExposure = { [weak self] (exposure, date) in self?.storeDailyExposure(exposure, at: date) }
```

nilsgrabenhorst commented on CoronaContact/Extensions/ENExposureDetectionSummary+totalRiskScore.swift:
```diff
@@ -36,18 +36,19 @@ private struct UniqueAttenuationLevelValues {
-    mutating func unique() -> UniqueAttenuationLevelValues? {
-        remove(0)
+    func unique() -> UniqueAttenuationLevelValues? {
+        var copy = Set(self)
```
`self` is already a Set.

finnsch committed:
CTAA-1389: Wrap call to a debugging method in a compiler directive

finnsch committed:
CTAA-1389: Implement suggestions from the code review
